scripts/genpinctrl/config.yaml

```diff
@@ -0,0 +1,158 @@
+    stm32mp1: 0x50002000
+    stm32wb: 0x48000000
+
+alternate-functions:
+    - name: ADC_IN / ADC_INN / ADC_INP
+      match: "^ADC(?:\\d+)?_IN[NP]?\\d+$"
+      mode: analog
+
+    - name: CAN_RX
+      match: "^CAN\\d+_RX$"
+      bias: pull-up
+      mode: input
+
+    - name: CAN_TX
+      match: "^CAN\\d+_TX$"
+      mode: output
+
+    - name: DAC_OUT
+      match: "^DAC(?:\\d+)?_OUT\\d+$"
+      mode: analog
+
+    - name: FDCAN_RX
+      match: "^FDCAN\\d+_RX$"
+      bias: pull-up
+      mode: input
+
+    - name: FDCAN_TX
+      match: "^FDCAN\\d+_TX$"
+      mode: output
+
+    - name: I2C_SCL
+      match: "^I2C\\d+_SCL$"
+      drive: open-drain
+      bias: pull-up
+      mode: output
+
+    - name: I2C_SDA
+      match: "^I2C\\d+_SDA$"
+      drive: open-drain
+      bias: pull-up
+      mode: inout
+
+    - name: I2S_CK
+      match: "^I2S\\d+_CK$"
+      mode: output
+
+    - name: I2S_WS
+      match: "^I2S\\d+_WS$"
+      mode: output
+
+    - name: I2S_SD
+      match: "^I2S\\d+_SD$"
+      mode: inout
+
+    - name: SPI_MASTER_MISO
+      match: "^SPI\\d+_MISO$"
+      bias: pull-down
+      mode: input
+      variant: master
+
+    - name: SPI_MASTER_MOSI
+      match: "^SPI\\d+_MOSI$"
+      bias: pull-down
+      mode: output
+      variant: master
+
+    - name: SPI_MASTER_SCK
+      match: "^SPI\\d+_SCK$"
+      bias: pull-down
+      mode: output
+      variant: master
+
+    - name: SPI_MASTER_NSS
+      match: "^SPI\\d+_NSS$"
+      bias: pull-up
+      mode: output
+      variant: master
+
+    - name: SPI_SLAVE_MISO
+      match: "^SPI\\d+_MISO$"
+      bias: pull-down
+      mode: output
+      variant: slave
+
+    - name: SPI_SLAVE_MOSI
+      match: "^SPI\\d+_MOSI$"
+      bias: pull-down
+      mode: input
+      variant: slave
+
+    - name: SPI_SLAVE_SCK
+      match: "^SPI\\d+_SCK$"
+      bias: pull-down
+      mode: input
+      variant: slave
+
+    - name: SPI_SLAVE_NSS
+      match: "^SPI\\d+_NSS$"
+      bias: pull-up
+      mode: input
+      variant: slave
+
+    - name: TIM_CH / TIM_CHN
+      match: "^TIM\\d+_CH\\d+N?$"
+      mode: output
+
+    - name: TIM_BKIN
+      match: "^TIM\\d+_BKIN\\d?$"
+      bias: pull-up
+      mode: input
+
+    - name: TIM_ETR
+      match: "^TIM\\d+_ETR$"
+      mode: input
+
+    - name: UART_CTS / USART_CTS / LPUART_CTS
+      match: "^(?:LP)?US?ART\\d+_CTS$"
+      drive: open-drain
+      bias: pull-up
+      mode: input
+
+    - name: UART_RTS / USART_RTS / LPUART_RTS
+      match: "^(?:LP)?US?ART\\d+_RTS$"
+      drive: open-drain
+      bias: pull-up
+      mode: output
+
+    - name: UART_TX / USART_TX / LPUART_TX
+      match: "^(?:LP)?US?ART\\d+_TX$"
+      bias: pull-up
+      mode: output
+
+    - name: UART_RX / USART_RX / LPUART_RX
+      match: "^(?:LP)?US?ART\\d+_RX$"
+      mode: input
+
+    - name: USB_OTG_FS_DM
+      match: "^USB_OTG_FS_DM$"
+      mode: inout
+
+    - name: USB_OTG_FS_DP
+      match: "^USB_OTG_FS_DP$"
+      mode: inout
```

Comment:
It is not clear what `inout` maps to within proposed dts bindings.

Reply:
The purpose of the `mode` field is to specify if pin is operating as `analog`, `input`, `output` or `inout` (input/output, like for example I2C). This field is only relevant to F1 series, where the pin mode is configured using the following rules:

- Analog: Analog mode (ANALOG)
- Alternate function input: Input mode (GPIO_IN)
- Alternate function output: Alternate Function Output (ALTERNATE)
- Alternate function input/output: Alternate Function Output (ALTERNATE)

It's basically needed for cases like `UART_RX`. Even though an alternate function is used, pin needs to be configured as `GPIO_IN`. This is what I understood from F1 datasheet.

Reply:
Ok if this field is f1 only. So maybe for clarity it could be renamed as `f1-mode` or alike.

Though, on F1 there is no input/output mode. From §9.1 'GPIO functional description'
![Screenshot from 2020-09-21 10-02-47](https://user-images.githubusercontent.com/18574308/93752198-95b2a800-fbfe-11ea-9fc7-4c16f519e9fd.png)
So, looking to proposed mapping from an external point of  view:
`analog` mode also translates to `ANALOG` (ok)
`input` mode translates to `GPIO_IN` (ok)
`output` mode translates to `ALTERNATE` (why not using `alternate` then.. but ok)
`inout` mode also translates to `ALTERNATE` (wait!?)
And we have no way to select `General Purpose output`, which I agree is not stricly needed, but it is available in dts binding and then we can get easily confused with the `ouput` one.

So at the end, I really consider that the mixing of F1 and other series makes things unclear.

Taking the example of UART_TX, the config.yaml proposes:
```
    - name: UART_TX / USART_TX / LPUART_TX
      match: "^(?:LP)?US?ART\\d+_TX$"
      bias: pull-up
      mode: output
```
For STM32F1, Reference manual mentions it should be "Alternate function push-pull".
It also adds: "The weak pull-up and pull-down resistors are disabled.", but looking at the config:
-`      mode: output`
 F1 only, but `output` translate to `ALTERNATE` in the end.
-`      bias: pull-up`
`pull-up` applies for all series, but for that case on F1, it doesn't apply (why it would have applied on an input mode for instance).

So at the end, since we're mixing series,
- on F1, it would happen that `bias` is specified, but not applied 
- for the F1 only `mode`, it translates to a different value (`output` > `alternate`)

I agree that, at the end, the correct configuration is generated.
But for someone looking the config from f1 point of view, it could seems quite unclear, or even buggy for someone that haven't read the reference manual.

If we're adding to that the fact that the default configuration are not mentioned (which is a choice that I totally support), this increases confusion to a higher level.

So in the end I really like the direction this is going to but this F1/other mixing is the dark point that we'd need to sort out. I agree this will have a cost of config file duplication, but will be much easier to understand for day to day and one time users.